lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -104,25 +256,97 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
     unsigned IdxForResult = 0;
     for (SILResultInfo RI : SubstitutedType->getIndirectResults()) {
       assert(RI.isIndirect());
-      if (RI.getSILType().isLoadable(M) && !RI.getType()->isVoid()) {
+      if (!RI.getSILType().hasTypeParameter() &&
+          RI.getSILType().isLoadable(M) && !RI.getType()->isVoid()) {
         Conversions.set(IdxForResult);
         break;
       }
       ++IdxForResult;
     }
   }
+
   // Try to convert indirect incoming parameters to direct parameters.
   unsigned IdxForParam = NumResults;
   for (SILParameterInfo PI : SubstitutedType->getParameters()) {
-    if (PI.getSILType().isLoadable(M) &&
+    if (!PI.getSILType().hasTypeParameter() &&
+        PI.getSILType().isLoadable(M) &&
         PI.getConvention() == ParameterConvention::Indirect_In) {
       Conversions.set(IdxForParam);
     }
     ++IdxForParam;
   }
+  // Produce a specialized type, which is the substituted type with
+  // the parameters/results passing conventions adjusted according
+  // to the converions selected above.
   SpecializedType = createSpecializedType(SubstitutedType, M);
 }
 
+/// Create a new substituted type with the updated signature.
+CanSILFunctionType
+ReabstractionInfo::createSubstitutedType(SILFunction *OrigF,
+                                         const TypeSubstitutionMap &SubstMap,
+                                         bool HasUnboundGenericParams) {
+  auto &M = OrigF->getModule();
+  auto SM = M.getSwiftModule();
+  auto OrigFnTy = OrigF->getLoweredFunctionType();
+
+  if (!HasUnboundGenericParams)
+    return SILType::substFuncType(M, SM, SubstMap, OrigFnTy,
+                                  /*dropGenerics = */ true);
+
+  // Explicitly subst generic types in the parameters
+  // and results.
+  // TODO: Can it be done in one go instead of doing it
+  // parameter-by-parameter?
+  SmallVector<SILParameterInfo, 8> SubstParams;
+  SubstParams.reserve(OrigFnTy->getParameters().size());
+  for (auto &OrigParam : OrigFnTy->getParameters()) {
+    auto NewTy = OrigParam.getType()
+                     .subst(SM, SubstMap, SubstFlags::AllowLoweredTypes)
+                     ->getCanonicalType();
+    if (NewTy->hasArchetype())
+      NewTy = OrigF->mapTypeOutOfContext(NewTy)->getCanonicalType();
+    if (!NewTy->isLegalSILType())
+      NewTy = M.Types.getLoweredType(NewTy).getSwiftRValueType();
+    SubstParams.push_back(SILParameterInfo(NewTy, OrigParam.getConvention()));
+  }
+
+  SmallVector<SILResultInfo, 8> SubstResults;
+  SubstResults.reserve(OrigFnTy->getNumAllResults());
+  for (auto OrigResult : OrigFnTy->getAllResults()) {
+    auto NewTy = OrigResult.getType()
+                     .subst(SM, SubstMap, SubstFlags::AllowLoweredTypes)
```

Comment:
You should be using SILType::subst() instead of AllowLoweredTypes -- I'd like to remove that flag at some point